### ISSUE TITLE
Tensorflow 2.5/2.4 ptxas and gpu issue

### ISSUE THREAD

**Issue by seermer** [tensorflow/tensorflow#50562]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow version: 2.4.0/2.4.2/2.5.0 gpu
- Python version: 3.8.0/3.8.10
- Installed using virtualenv? pip? conda?:  pip
- CUDA/cuDNN version: cuda 11.0/11.1/11.2.1/11.2.2, cudnn8.1.1

when running CNN, log shows:

> 2021-07-01 22:39:34.767676: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] Internal: ptxas exited with non-zero error code -1, output: 
> Relying on driver to perform ptx compilation. 
> Modify $PATH to customize ptxas location.
> This message will be only logged once.

the log also shows that it is **potentially using CPU and oneDNN**, even though gpu is available (and tf obviously can find it according to log)
the same message appears for combinations of tf2.4.0/2.4.2/2.5.0 with cuda11.0/11.1/11.2.1/11.2.2 (I tried several of these combinations, and the message always exist)
current combination on my machine: tf2.5.0+cuda11.2.1

**also tried: adding ptxas.exe path to system PATH**
doesn't help

similar to issue https://github.com/tensorflow/tensorflow/issues/49824 , but it seems that post is caused by signal handling, which I did not use

code:
```
import tensorflow as tf
import numpy as np

data = np.random.random((12800, 120, 120, 3,))
label = np.array([1 for _ in range(12800)])
print(data)

model = tf.keras.Sequential(
    [tf.keras.layers.Conv2D(256, kernel_size=5, strides=2, input_shape=(120, 120, 3,), activation="relu"),
     tf.keras.layers.Conv2D(256, kernel_size=5, strides=2, input_shape=(120, 120, 3,), activation="relu"),
     tf.keras.layers.GlobalAvgPool2D(),
     tf.keras.layers.Dense(1, activation="sigmoid")]
)
model.compile(loss=tf.keras.losses.BinaryCrossentropy())
model.fit(x=data, y=label, batch_size=32, epochs=20)
```
**note: this is just for demonstration, the same logging message appears whenever I use Conv2d in model**

full log:

> 2021-07-01 22:39:31.597866: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library nvcuda.dll
> 2021-07-01 22:39:31.623038: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
> pciBusID: 0000:01:00.0 name: NVIDIA GeForce GTX 1660 Ti computeCapability: 7.5
> coreClock: 1.59GHz coreCount: 24 deviceMemorySize: 6.00GiB deviceMemoryBandwidth: 268.26GiB/s
> 2021-07-01 22:39:31.623217: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cudart64_110.dll
> 2021-07-01 22:39:31.629267: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cublas64_11.dll
> 2021-07-01 22:39:31.629359: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cublasLt64_11.dll
> 2021-07-01 22:39:31.632662: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cufft64_10.dll
> 2021-07-01 22:39:31.633830: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library curand64_10.dll
> 2021-07-01 22:39:31.641547: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cusolver64_11.dll
> 2021-07-01 22:39:31.644421: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cusparse64_11.dll
> 2021-07-01 22:39:31.645145: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cudnn64_8.dll
> 2021-07-01 22:39:31.645317: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0
> 2021-07-01 22:39:31.645646: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX AVX2
> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
> 2021-07-01 22:39:31.646131: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
> pciBusID: 0000:01:00.0 name: NVIDIA GeForce GTX 1660 Ti computeCapability: 7.5
> coreClock: 1.59GHz coreCount: 24 deviceMemorySize: 6.00GiB deviceMemoryBandwidth: 268.26GiB/s
> 2021-07-01 22:39:31.646377: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0
> 2021-07-01 22:39:32.111044: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
> 2021-07-01 22:39:32.111146: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 
> 2021-07-01 22:39:32.111202: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N 
> 2021-07-01 22:39:32.111410: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3983 MB memory) -> physical GPU (device: 0, name: NVIDIA GeForce GTX 1660 Ti, pci bus id: 0000:01:00.0, compute capability: 7.5)
> 2021-07-01 22:39:33.207902: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)
> Epoch 1/20
> 2021-07-01 22:39:33.624659: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cudnn64_8.dll
> 2021-07-01 22:39:34.113561: I tensorflow/stream_executor/cuda/cuda_dnn.cc:359] Loaded cuDNN version 8101
> 2021-07-01 22:39:34.767676: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] Internal: ptxas exited with non-zero error code -1, output: 
> Relying on driver to perform ptx compilation. 
> Modify $PATH to customize ptxas location.
> This message will be only logged once.
> 2021-07-01 22:39:34.848389: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cublas64_11.dll
> 2021-07-01 22:39:35.307315: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cublasLt64_11.dll
> 400/400 [==============================] - 27s 60ms/step - loss: 0.0016
> Epoch 2/20
> 400/400 [==============================] - 24s 60ms/step - loss: 2.3526e-08
> Epoch 3/20
> 400/400 [==============================] - 24s 60ms/step - loss: 2.3526e-08...

**Comment by UsharaniPagadala** [comment]
@seermer 

I was able to reproduce the issue reported here in tf2.5 with no errors.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/fbff8db5d4e1645ef9f63651111f9754/untitled127.ipynb#scrollTo=RlmP-up0p_-V).Thanks

**Comment by seermer** [comment]
> @seermer
> 
> I was able to reproduce the issue reported here in tf2.5 with no errors.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/fbff8db5d4e1645ef9f63651111f9754/untitled127.ipynb#scrollTo=RlmP-up0p_-V).Thanks

Thanks for helping. Just to clarify, I don't think there will be any error showing up, since the message shows up as a part of the TensorFlow logging information. Colab doesn't output the log, so we can't see it.

I opened this issue only because exited with -1 sounds like something problematic, but the solution provided by log (add ptxas to $PATH) does not help at all. So I would like to know how do I solve this problem. And if possible, can the log message show clearer indications on how to solve it? Considering the current message doesn't really tell anything that helps to debug.

**Comment by sanjoy** [comment]
That message itself should be benign, but looping in @cheshire to check if the message can be made more actionable to users not familiar with TF internals.

**Comment by cheshire** [comment]
> (add ptxas to $PATH) does not help at all

It really should (remove the error message, not other issues) at least on Linux-based platforms. Sorry we are really short on Windows expertise. Maybe we are missing the `.exe` postfix when looking for it.

> I opened this issue because exited with -1 sounds like something problematic, cuda is used 0% when running model according to Task Manager, the log also shows it is using oneDNN, none of these seems to indicate TF-GPU is running correctly

As @sanjoy pointed out above these are completely independent issues and not related to the ptxas issue. Closing this issue, feel free to reopen with more detailed information (e.g. if you believe the GPU is not utilized when it should be).

**Comment by seermer** [comment]
@sanjoy @cheshire Sorry, could you please reopen the issue? Recently, I noticed that this issue seems not benign.
It is true that the model still runs without breaking with GPU even though this message appears. However, this error causes ANY jit compiled script to break. for example, running:
```
import tensorflow as tf

@tf.function(jit_compile=True)
def recompiled_on_launch(a, b):
  return a + b

recompiled_on_launch(tf.ones([1, 10]), tf.ones([1, 10]))
recompiled_on_launch(tf.ones([1, 100]), tf.ones([1, 100]))
```
gives the following error log:

> 2021-07-28 02:18:43.364806: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cudart64_110.dll
> 2021-07-28 02:18:45.982671: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library nvcuda.dll
> 2021-07-28 02:18:46.012216: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
> pciBusID: 0000:01:00.0 name: NVIDIA GeForce GTX 1660 Ti computeCapability: 7.5
> coreClock: 1.59GHz coreCount: 24 deviceMemorySize: 6.00GiB deviceMemoryBandwidth: 268.26GiB/s
> 2021-07-28 02:18:46.012382: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cudart64_110.dll
> 2021-07-28 02:18:46.103818: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cublas64_11.dll
> 2021-07-28 02:18:46.103918: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cublasLt64_11.dll
> 2021-07-28 02:18:46.142653: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cufft64_10.dll
> 2021-07-28 02:18:46.153116: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library curand64_10.dll
> 2021-07-28 02:18:46.263635: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cusolver64_11.dll
> 2021-07-28 02:18:46.289993: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cusparse64_11.dll
> 2021-07-28 02:18:46.294695: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library cudnn64_8.dll
> 2021-07-28 02:18:46.294846: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0
> 2021-07-28 02:18:46.295205: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX AVX2
> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
> 2021-07-28 02:18:46.295917: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
> pciBusID: 0000:01:00.0 name: NVIDIA GeForce GTX 1660 Ti computeCapability: 7.5
> coreClock: 1.59GHz coreCount: 24 deviceMemorySize: 6.00GiB deviceMemoryBandwidth: 268.26GiB/s
> 2021-07-28 02:18:46.296217: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1871] Adding visible gpu devices: 0
> 2021-07-28 02:18:46.806727: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
> 2021-07-28 02:18:46.806834: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 
> 2021-07-28 02:18:46.806894: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N 
> 2021-07-28 02:18:46.807170: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3983 MB memory) -> physical GPU (device: 0, name: NVIDIA GeForce GTX 1660 Ti, pci bus id: 0000:01:00.0, compute capability: 7.5)
> 2021-07-28 02:18:46.973764: I tensorflow/compiler/xla/service/service.cc:169] XLA service 0x20a9ed051d0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
> 2021-07-28 02:18:46.973880: I tensorflow/compiler/xla/service/service.cc:177]   StreamExecutor device (0): NVIDIA GeForce GTX 1660 Ti, Compute Capability 7.5
> 2021-07-28 02:18:47.040371: F tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc:472] ptxas returned an error during compilation of ptx to sass: 'Internal: ptxas exited with non-zero error code -1, output: '  If the error message indicates that a file could not be written, please verify that sufficient filesystem space is provided.
> 
> Process finished with exit code -1073740791 (0xC0000409)

as it shows "Process finished with exit code -1073740791 (0xC0000409)" it quits without continue running.
I have again tried reinstalling Anaconda, cuda, cudnn, tensorflow-gpu 2.5.0 according to the exact version specified many times, this problem stays.
So I would say this is not benign, and the new error message still doesn't help at all.

**Comment by cheshire** [comment]
To be honest, we really don't have enough Windows expertise to help you debug the issue. It seems ptxas is not found. You could try to add some print statements/logging in asm_compiler.cc to see where it is looking for it. Is your ptxas.exe definitely in your $PATH?

**Comment by seermer** [comment]
Alright, thanks, maybe I will just keep trying. I am quite sure I have ptxas.exe in path, and also all the important cuda/cudnn folders. I also have tried logging info, but the error points to compiling a temp file, which disappears instantly. Maybe I will just keep trying, thanks for helping.

**Comment by seermer** [comment]
@cheshire Alright, thanks, maybe I will just keep trying. I am quite sure I have ptxas.exe in path, and also all the important cuda folders. I also have tried logging info, but the error points to compiling a temp file, which disappears instantly. Maybe I will just keep trying, and see if I can somehow get the tempfile.
Thanks for helping.
